tests/test_animation.py

```diff
@@ -58,6 +58,22 @@ class TestAnimation:
     file_size = 4127
     caption = "Test *animation*"
 
+    def test_extra_slots(self):
+        animation = Animation(
+            self.animation_file_id,
+            self.animation_file_unique_id,
+            self.height,
+            self.width,
+            self.duration,
+        )
+        members = inspect.getmembers(
+            Animation,
+            predicate=lambda a: not inspect.isroutine(a) and (inspect.ismemberdescriptor(a)),
+        )
+        for member in members:
+            val = getattr(animation, member[0], 'err')
+            assert False if val == 'err' else True, f"got extra slot '{member[0]}'"
```

Comment:
```suggestion
            assert val != 'err', f"got extra slot '{member[0]}'"
```

Cool, didn't know about custom error messages with assert :D


But why exactly does this test that `animation` has no members that are not in `__slots__`? I would expect something like `name not in cls.__slots__`

Reply:
> Cool, didn't know about custom error messages with assert :D

Yea, super helpful for debugging.

> But why exactly does this test that `animation` has no members that are not in `__slots__`? I would expect something like `name not in cls.__slots__`

Actually the `members` variable contains the slots itself, so the only way was to check for extra slots is to check if we got a value (by calling getattr). And python instantly raises an error if there is a slot _missing_.

Also passing a class instance in `members = inspect.getmembers(...)` wouldn't quite work as it gives us all instance, class variables in a (key, value) fashion, without telling us which is which.

Reply:
Okay, I see the problem that getting a list of "only instance attributes" is not easy. But I still don't quite understand why this test checks that there are no extra slots. Could you try and explain differently?

Reply:
Because having extra slots would just take more memory? Why would we want to have extra slots anyway? Only having the exact amount is optimal.

Reply:
Sorry, bad phrasing. I do want the test to make that there are no extra slots. I just don't see why it does what it promises. I.e. I don't understand the code.

Reply:
1) `Get list of slots.` (just realized that I don't need inspect for this, could just use `cls.__slots__`)
2) `Call getattr for each slot to check if its actually used or not`
3) `if getattr returned 'err' it means that slot wasn't used and is extra`

If you were to check for slots how would you do it exactly?

Reply:
Haha, now I get it, thanks :) 👍 I think was was confused in my first comment, `name not in __slots__` would be for testing if an attribute is *not* in `__slots__` but that doesn't make any sense …